### PR TITLE
performance: send active controls as a single string per node

### PR DESCRIPTION
**Pull request by bboreham**
Scope "controls" are buttons to do something like restart a container.
These controls can be turned on and off by the probe, e.g. depending on whether the container is running.

Previously this information was conveyed by a complicated data structure mapping each control name to a `Dead` boolean which had a last-modified timestamp.  All of which takes effort to send and merge.
This PR changes the representation to a single string saying which controls are active, e.g. `"pause;restart;stop"`. String values have a timestamp, so the latest value prevails in a merge.

I guess we have to update the plugins docs but haven't really looked.

This is a protocol change, requires a version bump.

### CODE REVIEW

File: probe/awsecs/reporter.go [hunk 1]
```diff
@@ -150,17 +150,18 @@ func (r Reporter) Tag(rpt report.Report) (report.Report, error) {
 		// Create all the services first
 		for serviceName, service := range ecsInfo.Services {
 			serviceID := report.MakeECSServiceNodeID(cluster, serviceName)
+			activeControls := []string{ScaleUp}
+			// Disable ScaleDown when only 1 task is desired, since
+			// scaling down to 0 would cause the service to disappear (#2085)
+			if service.DesiredCount < 1 {
```
Comment (fbarl) [inline probe/awsecs/reporter.go]: Why did you change the condition? :thinking: 

```suggestion
			if service.DesiredCount <= 1 {
```

File: probe/docker/container.go [hunk 2]
```diff
@@ -389,20 +389,18 @@ func (c *container) getBaseNode() report.Node {
...
-		AttachContainer:  {Dead: !running},
-		ExecContainer:    {Dead: !running},
-		StartContainer:   {Dead: !stopped},
-		RemoveContainer:  {Dead: !stopped},
+	switch {
+	case paused:
+		return []string{UnpauseContainer}
+	case c.container.State.Running:
+		return []string{RestartContainer, StopContainer, PauseContainer, AttachContainer, ExecContainer}
+	case !c.container.State.Running:
```
Comment (fbarl) [inline probe/docker/container.go]: Nit: why not like this?

```suggestion
	default:
```

If the reason is readability, I feel it's a bit odd to single out the Running condition from above as the real condition that applies here is `!paused && !running`.

File: probe/docker/container.go [hunk 3]
```diff
@@ -389,20 +389,18 @@ func (c *container) getBaseNode() report.Node {
...
-		UnpauseContainer: {Dead: !paused},
-		RestartContainer: {Dead: !running},
-		StopContainer:    {Dead: !running},
-		PauseContainer:   {Dead: !running},
-		AttachContainer:  {Dead: !running},
-		ExecContainer:    {Dead: !running},
-		StartContainer:   {Dead: !stopped},
-		RemoveContainer:  {Dead: !stopped},
+	switch {
+	case paused:
```
Comment (fbarl) [inline probe/docker/container.go]: Nit: I guess I do have a little bit of OCD:

```suggestion
	case c.container.State.Paused:
```

File: report/node.go [hunk 4]
```diff
@@ -118,28 +117,16 @@ func (n Node) WithAdjacent(a ...string) Node {
...
-	return n
+	return n.WithLatest(NodeActiveControls, mtime.Now(), strings.Join(cs, ScopeDelim))
 }
 
-// WithLatestControl produces a new Node with control added to it
-func (n Node) WithLatestControl(control string, ts time.Time, data NodeControlData) Node {
-	n.LatestControls = n.LatestControls.Set(control, ts, data)
-	return n
+// ActiveControls returns a string slice with the names of active controls.
+func (n Node) ActiveControls(cs ...string) []string {
```
Comment (fbarl) [inline report/node.go]: `cs` doesn't seem to be used in the function or am I missing something? Weird how linter didn't complain:

```suggestion
func (n Node) ActiveControls() []string {
```

File: report/backcompat.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	*n = in.Node
+	if len(in.LatestControls) > 0 {
+		// Convert the map into a delimited string
+		cs := make([]string, 0, len(in.LatestControls))
+		var ts time.Time
+		for name, v := range in.LatestControls {
+			if !v.Value.Dead {
+				cs = append(cs, name)
+				// Pull out the newest timestamp to use for the whole set
+				if ts.Before(v.Timestamp) {
```
Comment (fbarl) [inline report/backcompat.go]: Is this always true before `ts` is set? Because otherwise we should make sure it doesn't end up being `nil` I guess?

File: report/backcompat.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+import (
+	"strings"
+	"time"
+
+	"github.com/ugorji/go/codec"
+)
+
+// For backwards-compatibility with probes that sent a map of latestControls data
+type bcNode struct {
```
Comment (fbarl) [inline report/backcompat.go]:
```suggestion
type nodeWithLatestControls struct {
```

Just for greater clarity and also to be open for the time when "backwards-compatible nodes" might diverge along more than just one property.

Comment (fbarl) [reply report/backcompat.go]: Actually never mind, I misinterpreted the intention here and forgot that this lives in `backcompat.go` :)

File: report/marshal_test.go [hunk 7]
```diff
@@ -71,3 +73,51 @@ func TestBiggerRoundtrip(t *testing.T) {
...
+        "latestControls": {
+          "docker_attach_container": {
+            "timestamp": "2019-10-14T14:36:01Z",
+            "value": {"dead": true}
+          },
+          "docker_remove_container": {
+            "timestamp": "2019-10-14T14:36:01Z",
+            "value": {"dead": false}
+          }
+        },
```
Comment (fbarl) [inline report/marshal_test.go]: Could you add a couple of more non-dead entries here? e.g. some with the same control name but different timestamps and some other with different control names for a more complete test?

File: report/marshal_test.go [hunk 8]
```diff
@@ -72,3 +73,51 @@ func TestBiggerRoundtrip(t *testing.T) {
...
+        "latestControls": {
+          "docker_attach_container": {
+            "timestamp": "2019-10-14T14:36:01Z",
+            "value": {"dead": true}
+          },
+          "docker_remove_container": {
+            "timestamp": "2019-10-14T14:36:01Z",
+            "value": {"dead": false}
+          }
```
Comment (fbarl) [inline report/marshal_test.go]: Could you add a couple more entries here? e.g. some with the same control name but different timestamps and some with differing control names for a more complete test?

File: report/backcompat.go (hunk 5 above)
Comment (bboreham) [reply report/backcompat.go]: `ts` is initialized to the "zero value" for `time.Time` which compares as before any real time.
`ts` cannot be `nil` as it is a value not a pointer type.

File: report/node.go (hunk 4 above)
Comment (bboreham) [reply report/node.go]: Right, it was bad copy/paste from the method above, thanks.
Go doesn't complain about unused parameters inside the function because they may be needed to conform to an interface.  And it's ok to supply an empty list when calling a `...` function.

File: report/marshal_test.go (hunk 8 above)
Comment (bboreham) [reply report/marshal_test.go]: I added a little bit more but not sure if I understand you.
